server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -195,6 +232,101 @@ public Status getStatus() {
+        }
+        final List<ThreadResourceInfo> threadResourceInfoList = resourceStats.computeIfAbsent(threadId, k -> new ArrayList<>());
+        // active thread entry should not be present in the list.
+        for (ThreadResourceInfo threadResourceInfo : threadResourceInfoList) {
+            if (threadResourceInfo.isActive()) {
+                throw new IllegalStateException("Unexpected active thread entry is present");
+            }
+        }
+        threadResourceInfoList.add(new ThreadResourceInfo(threadId, ResourceStatsType.WORKER_STATS, resourceUsageMetrics));
+    }
+
+    /**
+     * This method is used to update the resource consumption stats so that the data isn't too stale for long-running task.
+     * If an active thread entry is not present in the list, the update is dropped.
+     * @param threadId ID of the thread
+     * @param statsType stats type
+     * @param resourceUsageMetrics resource consumption metrics of the thread
+     */
+    public void updateTaskResourceStats(long threadId, ResourceStatsType statsType, ResourceUsageMetric... resourceUsageMetrics) {
+        final List<ThreadResourceInfo> threadResourceInfoList = resourceStats.get(threadId);
+        if (threadResourceInfoList == null) {
+            throw new IllegalStateException("Cannot update if thread resource info is not present");
+        } else {
+            // If active entry is not present, the update is dropped. If present, the active entry is updated.
+            for (ThreadResourceInfo threadResourceInfo : threadResourceInfoList) {
+                if (threadResourceInfo.isActive()) {
+                    threadResourceInfo.updateResourceInfo(statsType, resourceUsageMetrics);
+                    return;
+                }
+            }
+        }
+    }
+
+    /**
+     * Record the thread's end resource consumption values.
+     * @param threadId ID of the thread
+     * @param statsType stats type
+     * @param resourceUsageMetrics resource consumption metrics of the thread
+     */
+    public void stopTaskResourceStats(long threadId, ResourceStatsType statsType, ResourceUsageMetric... resourceUsageMetrics) {
+        final List<ThreadResourceInfo> threadResourceInfoList = resourceStats.get(threadId);
+        if (statsType != ResourceStatsType.WORKER_STATS || threadResourceInfoList == null) {
+            throw new IllegalArgumentException(
+                "Recording the end should have WORKER_STATS as stats type" + "and an active entry should be present in the list"
+            );
+        }
+        // marking active entries as done before updating the final resource usage values.
+        for (ThreadResourceInfo threadResourceInfo : threadResourceInfoList) {
+            if (threadResourceInfo.isActive()) {
+                threadResourceInfo.setActive(false);
+                threadResourceInfo.updateResourceInfo(ResourceStatsType.WORKER_STATS, resourceUsageMetrics);
+            }
+        }
+    }
+
```

Comment:
Same as above

Reply:
Will modify based on discussion on the other thread.